File: bindings/android/scion-http3-android/src/main/kotlin/com/anapaya/scion/http3/internal/ClientSettings.kt
// Copyright 2026 Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.anapaya.scion.http3.internal

import com.anapaya.scion.http3.PreferredUnderlay
import com.anapaya.scion.http3.SnapConfig
import com.anapaya.scion.http3.TrustAnchors
import com.anapaya.scion.http3.UdpConfig
import java.net.URI
import java.net.URISyntaxException

/**
 * Everything a client was configured with, validated, and holding no platform objects.
 *
 * Every setting except the endhost API and the trust anchors is nullable, and null means "whatever
 * the SCION stack does". That is the whole reason this type is separate from the FFI's own config:
 * a default restated here would be a second copy of a number the stack owns, and the two would
 * eventually disagree.
 *
 * Validation happens in [init], which is also what makes it testable: it needs no `Context`, so the
 * rejections can be exercised as ordinary unit tests rather than on a device.
 */
internal class ClientSettings(
    val endhostApiUrl: String,
    val trust: TrustAnchors,
    val authToken: String? = null,
    val preferredUnderlay: PreferredUnderlay? = null,
    val snap: SnapConfig? = null,
    val udp: UdpConfig? = null,
    val connectTimeoutMillis: Long? = null,
    val requestTimeoutMillis: Long? = null,
    val idleConnectionTimeoutMillis: Long? = null,
    val connectionAttemptDelayMillis: Long? = null,
    val maxOrigins: Int? = null,
    val maxResponseBodyBytes: Long? = null,
) {
    init {
        validateEndhostApiUrl(endhostApiUrl)
        authToken?.let { require(it.isNotEmpty()) { "an auth token cannot be empty" } }
        requirePositive("connectTimeout", connectTimeoutMillis)
        requirePositive("requestTimeout", requestTimeoutMillis)
        requirePositive("idleConnectionTimeout", idleConnectionTimeoutMillis)
        requirePositive("connectionAttemptDelay", connectionAttemptDelayMillis)
        maxOrigins?.let {
            require(it >= 1) { "maxOrigins has to be at least 1, got $it" }
        }
        maxResponseBodyBytes?.let {
            require(it > 0) { "maxResponseBody has to be positive, got $it bytes" }
        }
    }

    /**
     * How long a client may go unused before a request re-checks the network itself.
     *
     * Derived from the configured idle-connection timeout when there is one, since that is when the
     * connections a rebuild would discard have been swept anyway. With none configured this is a
     * heuristic of its own rather than a restatement of the stack's default: what matters is that
     * the gap is long enough not to fire between two ordinary requests, and short enough to catch a
     * change that happened while the application was in the background.
     */
    val idleThresholdMillis: Long
        get() =
            idleConnectionTimeoutMillis
                ?.let { StalenessTracker.idleThresholdFor(it) }
                ?: DEFAULT_IDLE_THRESHOLD_MILLIS

    private companion object {
        const val DEFAULT_IDLE_THRESHOLD_MILLIS: Long = 50_000

        fun requirePositive(
            name: String,
            millis: Long?,
        ) {
            millis?.let { require(it > 0) { "$name has to be positive, got $it ms" } }
        }

        fun validateEndhostApiUrl(url: String) {
            require(url.isNotEmpty()) {
                "endhostApi is required: it is the address the client discovers SCION " +
                    "connectivity through. A local PocketSCION topology reached from the " +
                    "emulator is http://10.0.2.2:8041."
            }
            val uri =
                try {
                    URI(url)
                } catch (e: URISyntaxException) {
                    throw IllegalArgumentException("endhostApi \"$url\" is not a valid URL", e)
                }
            require(uri.isAbsolute && !uri.host.isNullOrEmpty()) {
                "endhostApi \"$url\" needs a scheme and a host, for example " +
                    "https://endhost-api.example.org"
            }
            val scheme = uri.scheme.lowercase()
            require(scheme == "http" || scheme == "https") {
                "endhostApi \"$url\" has to be http or https, not $scheme"
            }
        }
    }
}

/**
 * Says, loudly, that certificate verification is off.
 *
 * Logged rather than refused, deliberately: a test against a throwaway server is a real need, and
 * the debuggable flag cannot tell an internal beta from a shipped app. So this makes noise instead
 * of decisions, and says plainly when the build it is running in is not a debug one.
 */
internal fun warnIfVerificationDisabled(
    trust: TrustAnchors,
    debug: DebugGuard,
    log: LibraryLog,
) {
    if (trust !is TrustAnchors.NoVerification) return
    log.error(
        "TLS certificate verification is DISABLED for this ScionHttp3Client. Every response " +
            "could come from anyone on the path. This is for local testing only.",
    )
    if (!debug.isDebuggable()) {
        log.error(
            "Worse: this application is not debuggable, so this is a release build. " +
                "TrustAnchors.insecureNoVerify() must not ship. Use TrustAnchors.pinned() with " +
                "the deployment's own certificate authority instead.",
        )
    }
}

File: bindings/apple/scion-http3-swift/Sources/ScionHTTP3/Internal/ClientSettings.swift
// Copyright 2026 Anapaya Systems
import Foundation

/// Everything a client was configured with.
///
/// Every setting except the endhost API and the trust anchors is optional and nil means "whatever
/// the SCION stack does".
struct ClientSettings: Sendable, Equatable {
    let endhostApiUrl: String
    let authToken: String?
    let preferredUnderlay: PreferredUnderlay?
    let snap: SnapConfig
    let udp: UdpConfig
    let trust: TrustAnchors
    let connectTimeout: TimeInterval?
    let requestTimeout: TimeInterval?
    let idleConnectionTimeout: TimeInterval?
    let connectionAttemptDelay: TimeInterval?
    let maxOrigins: Int?
    let maxResponseBodyBytes: Int?

    /// Throws `ScionHttp3Error.invalidConfiguration` for the first setting that cannot be right.
    init(_ configuration: ScionHttp3Client.Configuration) throws {
        try Self.validateEndhostApiUrl(configuration.endhostApi)
        if let token = configuration.authToken, token.isEmpty {
            throw ScionHttp3Error.invalidConfiguration(detail: "an auth token cannot be empty")
        }
        try Self.requirePositive("connectTimeout", configuration.connectTimeout)
        try Self.requirePositive("requestTimeout", configuration.requestTimeout)
        try Self.requirePositive("idleConnectionTimeout", configuration.idleConnectionTimeout)
        try Self.requirePositive("connectionAttemptDelay", configuration.connectionAttemptDelay)
        if let maxOrigins = configuration.maxOrigins, maxOrigins < 1 {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "maxOrigins has to be at least 1, got \(maxOrigins)")
        }
        if let bytes = configuration.maxResponseBodyBytes, bytes <= 0 {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "maxResponseBodyBytes has to be positive, got \(bytes) bytes")
        }
        if let index = configuration.snap.dataPlaneIndex, index < 0 {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "a data plane index cannot be negative, got \(index)")
        }
        if let key = configuration.snap.staticIdentity, key.count != 32 {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "a SNAP static identity is a 32-byte X25519 private key, "
                    + "got \(key.count) bytes")
        }
        if configuration.udp.outboundIps.contains(where: {
            $0.trimmingCharacters(in: .whitespaces).isEmpty
        }) {
            throw ScionHttp3Error.invalidConfiguration(detail: "an outbound IP cannot be blank")
        }
        try Self.requirePositive(
            "nextHopResolverFetchInterval", configuration.udp.nextHopResolverFetchInterval)

        endhostApiUrl = configuration.endhostApi
        authToken = configuration.authToken
        preferredUnderlay = configuration.preferredUnderlay
        snap = configuration.snap
        udp = configuration.udp
        trust = configuration.trust
        connectTimeout = configuration.connectTimeout
        requestTimeout = configuration.requestTimeout
        idleConnectionTimeout = configuration.idleConnectionTimeout
        connectionAttemptDelay = configuration.connectionAttemptDelay
        maxOrigins = configuration.maxOrigins
        maxResponseBodyBytes = configuration.maxResponseBodyBytes
    }

    /// How long a client may go unused before a request re-checks the network itself.
    var idleThreshold: TimeInterval {
        idleConnectionTimeout.map(StalenessTracker.idleThreshold(forIdleConnectionTimeout:))
            ?? Self.defaultIdleThreshold
    }

    static let defaultIdleThreshold: TimeInterval = 50

    private static func requirePositive(_ name: String, _ value: TimeInterval?) throws {
        if let value, !(value > 0) || !value.isFinite {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "\(name) has to be positive and finite, got \(value) s")
        }
    }

    private static func validateEndhostApiUrl(_ url: String) throws {
        if url.isEmpty {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "endhostApi is required: it is the address the client discovers SCION "
                    + "connectivity through. A local PocketSCION topology is reached at the "
                    + "endhost API URL it prints, for example http://127.0.0.1:8041.")
        }
        guard let components = URLComponents(string: url) else {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "endhostApi \"\(url)\" is not a valid URL")
        }
        guard let scheme = components.scheme, let host = components.host, !host.isEmpty else {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "endhostApi \"\(url)\" needs a scheme and a host, for example "
                    + "https://endhost-api.example.org")
        }
        let lowered = scheme.lowercased()
        if lowered != "http" && lowered != "https" {
            throw ScionHttp3Error.invalidConfiguration(
                detail: "endhostApi \"\(url)\" has to be http or https, not \(scheme)")
        }
    }
}

func warnIfVerificationDisabled(_ trust: TrustAnchors, log: any LibraryLog) {
    guard trust.kind == .insecureNoVerify else { return }
    log.error(
        "TLS certificate verification is DISABLED for this ScionHttp3Client. Every response could "
            + "come from anyone on the path. This is for local testing only. "
            + "TrustAnchors.insecureNoVerify should not ship. Use TrustAnchors.pinned(_:) with the "
            + "deployment's own certificate authority instead.")
}
